assets/js/util/when-active.js

```diff
@@ -0,0 +1,65 @@
+/**
+ * whenActive HOC.
+ *
+ * Site Kit by Google, Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * WordPress dependencies
+ */
+import { createElement } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import Data from 'googlesitekit-data';
+import { STORE_NAME as MODULES_STORE } from '../googlesitekit/modules/datastore/constants';
+import { kebabCaseToPascalCase } from '../googlesitekit/data/transform-case';
+const { useSelect } = Data;
+
+/**
+ * Higher-Order Component to render wrapped components when selected module is active and connected.
+ *
+ * @since n.e.x.t
+ *
+ * @param {Object}   options                     Options for enhancing function.
+ * @param {string}   options.moduleName          Name of a module to check.
+ * @param {Function} [options.fallbackComponent] Optional. Fallback component to render when the module is not active.
+ * @return {Function} Enhancing function.
+ */
+export default function whenActive( { moduleName, fallbackComponent = null } ) {
+	return ( wrappedComponent ) => {
+		const whenActiveComponent = ( props ) => {
+			// eslint-disable-next-line react-hooks/rules-of-hooks
+			const moduleInfo = useSelect( ( select ) => select( MODULES_STORE ).getModule( moduleName ) );
```

Comment:
Why is this being disabled here? This component is getting created every time this code is run, so this points to it being better to create this as its own component then include it inside the HOC, if we stick with the HOC approach...

Reply:
It's disabled here because that eslint rule incorrectly detects the current use case for `useSelect` hook and thinks that it's called outside of a component which is not the case.

> This component is getting created every time this code is run, so this points to it being better to create this as its own component then include it inside the HOC, if we stick with the HOC approach...

That's exactly how it should work because we need to create a new component for every wrapping component.

Reply:
Perhaps it wouldn't error if using [`createHigherOrderComponent`](https://github.com/WordPress/gutenberg/blob/master/packages/compose/src/utils/create-higher-order-component/index.js)? WP uses several HOCs and we use their ruleset so it's a bit strange that it would raise this linting error.

Reply:
@aaemnnosttv unfortunately, it doesn't help.

Reply:
What about moving the `useSelect` above, outside the newly-created component. Technically, this would avoid the hooks violation and doesn't seem like a problem. Would that work? eg:

```js
export default function whenActive( { moduleName, fallbackComponent = null } ) {
	return ( wrappedComponent ) => {
		const isConnected = useSelect( ( select ) => select( CORE_MODULES ).isModuleConnected( moduleName ) );

		const whenActiveComponent = ( props ) => {
			// Return null if the module is not loaded yet or doesn't exist.
			if ( typeof isConnected === 'undefined' || isConnected === null ) {
				return null;
			}

[...]
```

Reply:
@tofumatt no, unfortunately, we can't do it too, it has to be in the `whenActiveComponent` because if we move it outside, we will definitely break the "rules of hooks" then.